Lib/json/tool.py

```diff
@@ -69,6 +69,10 @@ def main():
             if options.outfile is None:
                 out = sys.stdout
             else:
+                if (options.outfile.exists() and
+                    options.outfile.expanduser().samefile(infile.name)):
```

Comment:
There is no guarantee that samefile() works as expected for all cases.
For example,

```
$ ./python.exe x.py | ./python.exe -m json.tool --json-lines - out.jsonl
Traceback (most recent call last):
  File "/Users/inada-n/work/python/cpython/Lib/pathlib.py", line 1006, in samefile
    other_st = other_path.stat()
               ^^^^^^^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'stat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/inada-n/work/python/cpython/Lib/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/inada-n/work/python/cpython/Lib/runpy.py", line 86, in _run_code
    exec(code, run_globals)
    ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/inada-n/work/python/cpython/Lib/json/tool.py", line 87, in <module>
    main()
    ^^^^^^
  File "/Users/inada-n/work/python/cpython/Lib/json/tool.py", line 73, in main
    options.outfile.expanduser().samefile(infile.name)):
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/inada-n/work/python/cpython/Lib/pathlib.py", line 1008, in samefile
    other_st = self._accessor.stat(other_path)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '<stdin>'
```

There may be other cases that `samefile()` don't work as your expect.

Additionally, if we don't load the whole file at once, we shouldn't do that even when infile and outfile is same file.

So I think creating a temporary file and rename on success is the right way.

Reply:
That's rude, github... I had written a response and pressed "Resolve" and it didn't post my comment. Anyway the gist was:

[...]
> There may be other cases that samefile() don't work as your expect.

Thanks for the counterexample the `samefile()` docs clearly link to the `os.stat` cases, I should have followed them.

> Additionally, if we don't load the whole file at once, we shouldn't do that even when infile and outfile is same file.
> 
> So I think creating a temporary file and rename on success is the right way.

And while I think the tempfile idiom/technique has valid applications, the aim I had with  #29273 was to implement what `sponge` and `sort -o` do, with little added complexity. Having seen the mess tools like pip and pipenv make of my /tmp, I feel uncomfortable introducing it into a small tool like json.tool.

So like you said
> json.tool is a toy/demo program. So I don't think we need to support every case.

I've closed this PR. I'm unsure what should happen with the bpo. Feels a bit bad to just scratch my own itch and then drop the ball, but this rabbit hole goes a bit deeper than I initially judged. Thanks again for your time and patience.